src/Mono.Android/Android.Runtime/JNIEnv.cs

```diff
@@ -632,13 +638,49 @@ public static string GetClassNameFromInstance (IntPtr jobject)
 		}
 
 		[DllImport ("__Internal", CallingConvention = CallingConvention.Cdecl)]
-		internal static extern IntPtr monodroid_typemap_managed_to_java (string managed);
+		static extern IntPtr monodroid_typemap_managed_to_java (byte[] mvid, int token);
+
+		internal static void LogTypemapTrace (StackTrace st)
+		{
+			string trace = st.ToString ()?.Trim ();
+			if (String.IsNullOrEmpty (trace))
+				return;
+
+			monodroid_log (LogLevel.Warn, LogCategories.Assembly, "typemap: called from");
+			foreach (string line in trace.Split ('\n')) {
+				monodroid_log (LogLevel.Warn, LogCategories.Assembly, line);
+			}
+		}
+
+		internal static IntPtr TypemapManagedToJava (Type type)
```

Comment:
I half wonder if this method should return `string` instead of `IntPtr`, since every call site calls `Marshal.PtrToStringAnsi()` *anyway* on non-null return values...